[COMSUP-154] Edge functions caching (#903)
edgio-docs/edgio-docs: src/guides/v7/edge_functions.md

}
```

### Caching fetch() Requests {/* caching-fetch-requests */}

> Caching fetch requests within your edge function can reduce the load on your origins and deliver content faster to your users. It's important to be aware of [time limitations](#limitations) of edge function which can also be mitigated with caching.

Reviewer: Consider:
```suggestion
Caching fetch requests within your edge function can reduce the load on your origins and deliver content faster to your users. It may also mitigate timeout issues due to an edge function exceeding the [walltime limit](#limitations).
```

}
```

## Caching fetch() Requests {/* caching-fetch-requests */}

Caching fetch requests within your edge function can reduce the load on your origins and deliver content faster to your users. It may also mitigate timeout issues due to an edge function exceeding the [walltime limit](#limitations).

In this section, we'll cover how to use the caching options as part of the `fetch()` method. These options are specified per fetch request and are completely separate from the caching options specified in {{ROUTES_FILE}}.

The following sample code shows how to define caching options inside the `edgio` object:

```js
const resp = await fetch('https://your-server.com/some-path', {
  edgio: {
    origin: 'web',
    caching: {
      max_age: '1d',
      stale_while_revalidate: '1h',
      tags: 'apple banana',
      bypass_cache: false,

>     },

Reviewer 2: This is not available right now. It will probably be available at the end of November.

- `stale_while_revalidate`: Specifies the amount of time a stale response is served while a revalidation request is made in the background. This value is also set as a duration string similar to `max_age`. This setting overrides the `stale-while-revalidate` directive in the `Cache-Control` header of the origin response if present.
- `tags`: Allows you to specify a space-separated list of tags for the cached object, which can later be used for cache purging as [surrrogate keys](/guides/performance/caching/purging#surrogate-key). Each tag should be a string without spaces.
- `bypass_cache`: A boolean value that, when set to `true`, bypasses the cache for the fetch request, ensuring the request is sent directly to the origin and the response is not stored in the cache.

> These caching options provide you with granular control over how your fetch requests are cached and served, allowing you to optimize the performance of your edge function.

Reviewer 2: Not available right now.

### Cache Behavior

> Edge function subrequests are cached at the edge for 5 minutes by default. This means that if you make a fetch request to the same URL within 5 minutes, the response will be served from the cache instead of going to the origin. This behavior can be overridden by specifying the `bypass_cache` option. Cache directives from the origin response will also be respected as follows:

Reviewer 2: > Edge function subrequests are cached at the edge for 5 minutes by default

This statement is too broad. We only cache requests for 5 minutes that don't have explicit `Cache-Control` response headers **and** that are considered "cachable".

If a response has valid `Cache-Control` headers, they will be respected.

- `stale_while_revalidate`: Specifies the amount of time a stale response is served while a revalidation request is made in the background. This value is also set as a duration string similar to `max_age`. This setting overrides the `stale-while-revalidate` directive in the `Cache-Control` header of the origin response if present.
- `tags`: Allows you to specify a space-separated list of tags for the cached object, which can later be used for cache purging as [surrrogate keys](/guides/performance/caching/purging#surrogate-key). Each tag should be a string without spaces.
- `bypass_cache`: A boolean value that, when set to `true`, bypasses the cache for the fetch request, ensuring the request is sent directly to the origin and the response is not stored in the cache.

These caching options provide you with granular control over how your fetch requests are cached and served, allowing you to optimize the performance of your edge function. -->

### Cache Behavior of Subrequests {/* cache-behavior-of-subrequests */}

Edge function subrequests are cached at the edge for 5 minutes under the following conditions:

> - The origin does not offer a `Cache-Control` header in the response.

Reviewer: ```suggestion
- The response from the origin does not include a `Cache-Control` header.
```

- The response is deemed cacheable based on our [default caching policy](/guides/performance/caching#default-caching-policy).

This means that if you make a fetch request to the same URL within 5 minutes, the response will be served from the cache instead of going to the origin. <!-- This behavior can be overridden by specifying the `bypass_cache` option. --> Cache directives from the origin response will also be respected as follows:

- If the origin responds with a `Cache-Control` HTTP header containing valid directives, these directives will be respected. For example:
  - With `Cache-Control: max-age=60, s-maxage=900`, the fetch request will be cached for 15 minutes, considering `s-maxage=900`.
  - With `Cache-Control: max-age=600`, the response will be cached for 10 minutes, considering `max-age=600`.
  - With `Cache-Control: no-store, no-cache`, the response will not be cached.
- If the `Cache-Control` header is not present for a cache-eligible response, the CDN will check for the `Expires` header.

> - If the CDN determines that the response can be cached, it will be cached for 5 minutes on the edge.

Author: ```suggestion
```

> - If the response is cached based on the above logic, subsequent fetch requests will fetch from the cache, and only go to the origin if the response has expired or doesn't exist in the cache.

Reviewer: ```suggestion
- If the response is cached based on the above logic, subsequent fetch requests will be served from cache until the cached response has expired or been purged. At which point, the fetch request will go to the origin.
```
